applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h

```diff
@@ -0,0 +1,274 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Miguel Maso Sotomayor
+//
+
+#ifndef KRATOS_REPLICATE_MODEL_PART_UTILITY
+#define KRATOS_REPLICATE_MODEL_PART_UTILITY
+
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+#include "includes/model_part.h"
+
+
+namespace Kratos
+{
+///@addtogroup ShallowWaterApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+*/
+class KRATOS_API(SHALLOW_WATER_APPLICATION) ReplicateModelPartUtility
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    typedef std::size_t IndexType;
+
+    /// Pointer definition of ReplicateModelPartUtility
+    KRATOS_CLASS_POINTER_DEFINITION(ReplicateModelPartUtility);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+    ReplicateModelPartUtility(ModelPart& rOriginModelPart, ModelPart& rDestinationModelPart, bool ReplicateSubModelParts = false);
+
+    /// Destructor.
+    ~ReplicateModelPartUtility() = default;
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void Replicate();
+
+    template<class TVarType>
+    void TransferVariable(TVarType& rVariable)
+    {
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(mrOriginModelPart.NumberOfNodes()); ++i)
+        {
+            const auto it_node = mrOriginModelPart.NodesBegin() + i;
+            const auto dest_node = mReplicatedNodes[it_node->Id()];
+            dest_node->FastGetSolutionStepValue(rVariable) = it_node->FastGetSolutionStepValue(rVariable);
+        }
+    }
+
+    template<class TVarType>
+    void TransferNonHistoricalVariable(TVarType& rVariable)
```

Comment:
I think that there is something in the `VariableUtils` that does this operation. Have you check?